File: netlogo-gui/src/main/agent/InRadiusOrCone3D.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.api.{ AgentException, Dump, Vect }
import org.nlogo.core.AgentKind

import java.lang.{ Long => JLong }
import java.util.{ ArrayList => JArrayList, HashSet => JHashSet, Iterator => JIterator, List => JList, RandomAccess }

import scala.collection.mutable.ArrayBuffer

object InRadiusOrCone3D {
  trait ThreeDRadialAgent[A <: Agent] extends InRadiusOrCone.RadialAgent[A] {
    def agentKind: AgentKind
    def x(a: A): Double
    def y(a: A): Double
    def z(a: A): Double
    def globalSet(w: World): AgentSet
    def foreachAgent(p: Patch, f: A => Unit): Unit
    def estimatedAgentCount(patchCount: Int, world: World): Long
  }

  implicit object ThreeDRadialPatch extends ThreeDRadialAgent[Patch3D] {
    def agentKind: AgentKind = AgentKind.Patch
    def x(a: Patch3D): Double = a.pxcor
    def y(a: Patch3D): Double = a.pycor
    def z(a: Patch3D): Double = a.pzcor
    def globalSet(w: World): AgentSet = w.patches
    def foreachAgent(p: Patch, f: Patch3D => Unit): Unit = f(p.asInstanceOf[Patch3D])
    def estimatedAgentCount(patchCount: Int, world: World): Long = patchCount
  }

  implicit object ThreeDRadialTurtle extends ThreeDRadialAgent[Turtle3D] {
    def agentKind: AgentKind = AgentKind.Turtle
    def x(a: Turtle3D): Double = a.xcor
    def y(a: Turtle3D): Double = a.ycor
    def z(a: Turtle3D): Double = a.zcor
    def globalSet(w: World): AgentSet = w.patches
    def foreachAgent(p: Patch, f: Turtle3D => Unit): Unit = {
      p.turtlesHere match {
        // this optimization may seem minor, but actually results in ~ 5-10% speedup
        // when finding any turtles in-radius
        case arrayList: JArrayList[_] =>
          var i = 0
          while (i < arrayList.size) {
            f(arrayList.get(i).asInstanceOf[Turtle3D])
            i += 1
          }
        case turtles =>
          val turtleIterator = turtles.iterator
          while (turtleIterator.hasNext) {
            f(turtleIterator.next().asInstanceOf[Turtle3D])
          }
      }
    }
    def estimatedAgentCount(patchCount: Int, world: World): Long =
      patchCount * (world.turtles.sizeBound / world.patches.sizeBound)
  }
}

import InRadiusOrCone._
import InRadiusOrCone3D._

class InRadiusOrCone3D(world3D: World3D) extends InRadiusOrCone(world3D) {

  override def generateRadiusStrategies(
    agent: Agent,
    sourceSet: AgentSet,
    radius: Double,
    wrap: Boolean): Seq[Strategy] = {
    val (startPatch, startX, startY, startZ) =
      agent match {
        case startTurtle: Turtle3D =>
          (startTurtle.getPatchHere.asInstanceOf[Patch3D], startTurtle.xcor(), startTurtle.ycor(), startTurtle.zcor)
        case patch: Patch3D =>
          (patch, patch.pxcor.toDouble, patch.pycor.toDouble, patch.pzcor.toDouble)
        case a =>
          throw new IllegalStateException("asked to find in-radius on unexpected agent: " + a)
      }

    val sourceKind = sourceSet.kind

    if (sourceKind == AgentKind.Patch) {
      val inRadius = new VerifyInRadius[Patch3D](startX, startY, startZ, radius, wrap)
      sharedStrategies[Patch3D](inRadius, sourceSet, world3D,
        new RadialPatchSetEnumeration(inRadius, _, radius, sourceSet, startPatch))
    } else if (sourceKind == AgentKind.Turtle) {
      val inRadius = new VerifyInRadius[Turtle3D](startX, startY, startZ, radius, wrap)
      val breededTurtles = new VerifyMembershipBreedset[Turtle3D](sourceSet, world3D)
      new RadialPatchSetEnumeration(inRadius, breededTurtles, radius, sourceSet, startPatch) +:
      sharedStrategies[Turtle3D](inRadius, sourceSet, world3D,
        new RadialPatchSetEnumeration(inRadius, _, radius, sourceSet, startPatch))
    } else
      Seq()
  }

  override def generateConeStrategies(
    callingTurtle: Turtle,
    sourceSet: AgentSet,
    radius: Double,
    angle: Double,
    wrap: Boolean): Seq[Strategy] = {
    val startTurtle = callingTurtle.asInstanceOf[Turtle3D]
    val startPatch = startTurtle.getPatchHere.asInstanceOf[Patch3D]

    val half = angle / 2.0

    if (sourceSet.kind == AgentKind.Patch) {
      val inCone = new VerifyInCone[Patch3D](startTurtle, radius, wrap, half)
      sharedStrategies[Patch3D](inCone, sourceSet, world3D,
        new ConicPatchSetEnumeration(inCone, _, radius, sourceSet, startPatch, world3D))
    } else {
      val inCone = new VerifyInCone[Turtle3D](startTurtle, radius, wrap, half)
      val breededTurtles = new VerifyMembershipBreedset[Turtle3D](sourceSet, world)
      new ConicPatchSetEnumeration(inCone, breededTurtles, radius, sourceSet, startPatch, world3D) +:
      sharedStrategies[Turtle3D](inCone, sourceSet, world3D,
        new ConicPatchSetEnumeration(inCone, _, radius, sourceSet, startPatch, world3D))
    }
  }

  private class VerifyInRadius[A <: Agent](startX: Double, startY: Double, startZ: Double, radius: Double, wrap: Boolean)(implicit pos: ThreeDRadialAgent[A]) extends VerifyGeometry[A] {
    val protractor = world.protractor.asInstanceOf[Protractor3D]

    def inGeometry(a: A) =
      protractor.distance(pos.x(a), pos.y(a), pos.z(a), startX, startY, startZ, wrap) <= radius
  }

  def sharedStrategies[A <: Agent](
    geometry: VerifyGeometry[A], sourceSet: AgentSet, world: World3D, f: VerifyMembership[A] => Strategy)(
      implicit ev: ThreeDRadialAgent[A]): Seq[Strategy] = {
    val allMembers       = new VerifyMembershipGlobal[A](world)
    val arbitraryMembers = new VerifyMembershipArbitrary[A](sourceSet)
    Seq(new ExhaustiveEnumeration[A](geometry, sourceSet), f(allMembers), f(arbitraryMembers))
  }

  private class VerifyInCone[A <: Agent](startTurtle: Turtle3D, radius: Double, wrap: Boolean, half: Double)(implicit pos: ThreeDRadialAgent[A]) extends VerifyGeometry[A] {
    val worldWidth = world.worldWidth
    val worldHeight = world.worldHeight
    val worldDepth = world3D.worldDepth
    val protractor = world3D.protractor.asInstanceOf[Protractor3D]
    val startX = startTurtle.xcor()
    val startY = startTurtle.ycor()
    val startZ = startTurtle.zcor()
    val turtleHeading = startTurtle.heading
    val turtlePitch = startTurtle.pitch

    // If wrap is true and the radius is large enough, the cone
    // wraps around the edges of the world.  We handle this by
    // re-attempting the in-cone search multiple times, offset by
    // the world width and height (as appropriate).
    // m and n are the maximum number of times the cone might wrap
    // around the edge of the world in the X and Y directions.
    // The searches are repeated for the values [-m, m] on the x axis,
    // [-n, n] on the y axis, [-k, k] on the z-axis.
    // So for each cone check we may perform (2n * 2m * 2k) separate
    // cone checks with the geometry offset.
    val (m, n, k) =
      if (wrap) {
        (if (world.wrappingAllowedInX) StrictMath.ceil(radius / worldWidth).toInt else 0,
         if (world.wrappingAllowedInY) StrictMath.ceil(radius / worldHeight).toInt else 0,
         if (world3D.wrappingAllowedInZ) StrictMath.ceil(radius / worldDepth).toInt else 0)
      } else {
        (0, 0, 0)
      }

    def inGeometry(a: A): Boolean = {
      val agentX = pos.x(a)
      val agentY = pos.y(a)
      val agentZ = pos.z(a)
      // we want to start with worldOffset 0 and work our way "outward"
      var worldOffsetX = 0
      while (worldOffsetX <= m) {
        var worldOffsetY = 0
        while (worldOffsetY <= n) {
          var worldOffsetZ = 0
          while (worldOffsetZ <= k) {
            if (isInCone(agentX + worldWidth * worldOffsetX, agentY + worldHeight * worldOffsetY, agentZ + worldDepth * worldOffsetZ))
              return true
            worldOffsetZ =
              if (worldOffsetZ < 0) (- worldOffsetZ) + 1
              else if (worldOffsetZ > 0) - worldOffsetZ
              else worldOffsetZ + 1
          }
          worldOffsetY =
            if (worldOffsetY < 0) (- worldOffsetY) + 1
            else if (worldOffsetY > 0) - worldOffsetY
            else worldOffsetY + 1
        }
        worldOffsetX =
          if (worldOffsetX < 0) (- worldOffsetX) + 1
          else if (worldOffsetX > 0) - worldOffsetX
          else worldOffsetX + 1
      }
      false
    }

    def isInCone(x: Double, y: Double, z: Double): Boolean = {
        if (x == startX && y == startY && z == startZ) {
          return true;
        }
        if (protractor.distance(startX, startY, startZ, x, y, z, false) > radius) // false = don't wrap, since inCone()
          // handles wrapping its own way
          {
            return false;
          }

          val unitVect = Vect.toVectors(turtleHeading, turtlePitch, 0)(0)
          val targetVect = new Vect(x - startX, y - startY, z - startZ)
          val angle = targetVect.angleTo(unitVect)
          val halfRadians = StrictMath.toRadians(half)

          return angle <= halfRadians || angle >= (2 * StrictMath.PI - halfRadians);
      }
  }

  private abstract class PatchSetEnumeration[A <: Agent](
    geometry: VerifyGeometry[A],
    membership: VerifyMembership[A],
    radius: Double,
    sourceSet: AgentSet,
    startPatch: Patch3D)(implicit ra: ThreeDRadialAgent[A]) extends Strategy {
    val name = s"Patch-based search on ${membership.name}"
    val rootsTable = world.rootsTable
    val ((dxmax, dxmin), (dymax, dymin), (dzmax, dzmin)) = computePatchRanges
    val numberOfPatches = (dxmax - dxmin) * (dymax - dymin) * (dzmax - dzmin)
    val estimatedAgentCount: Long = ra.estimatedAgentCount(numberOfPatches, world)
    def applies: Boolean = membership.appliesTo(sourceSet)
    def cost: Long = membership.cost + (numberOfPatches * CostToGetPatchAtOffset) +
      (estimatedAgentCount * (geometry.costPerAgent + membership.costPerAgent))
    def getPatch(dx: Int, dy: Int, dz: Int): Patch3D // this is different for radius / cone
    def findAgents: IndexedAgentSet = {
      membership.beginInclusionChecks()
      val result = new AgentSetBuilder(ra.agentKind, StrictMath.min(estimatedAgentCount, membership.estimatedSize).toInt)
      var dz = dzmin
      while (dz <= dzmax) {
        var dy = dymin
        while (dy <= dymax) {
          var dx = dxmin
          while (dx <= dxmax) {
            try {
              val gridRoot = StrictMath.sqrt(dx * dx + dy * dy + dz * dz)
              if (gridRoot <= radius + 1.415) {
                val patch = getPatch(dx, dy, dz)
                ra.foreachAgent(patch, { (a: A) =>
                  if (membership.includes(a) && geometry.inGeometry(a))
                    result.add(a)
                })
                // it's possible to optimize this further (at the cost of abstraction)
                // by skipping the Geometry check for in-radius if gridRoot < radius - 1.415
              }
            } catch {
              case e: AgentException => org.nlogo.api.Exceptions.ignore(e)
            }
            dx += 1
          }
          dy += 1
        }
        dz += 1
      }
      result.build()
    }

    private def computePatchRanges: ((Int, Int), (Int, Int), (Int, Int)) = {
      val r = StrictMath.ceil(radius).toInt

      (computePatchRange(world.wrappingAllowedInX,    world.worldWidth,   world.minPxcor,
          world.maxPxcor,   startPatch.pxcor, r),
        computePatchRange(world.wrappingAllowedInY,   world.worldHeight,  world.minPycor,
          world.maxPycor,   startPatch.pycor, r),
        computePatchRange(world3D.wrappingAllowedInZ, world3D.worldDepth, world3D.minPzcor,
          world3D.maxPzcor, startPatch.pzcor, r))
    }

    // make sure to use half the world dimensions rather than just max-p(x/y)cor
    // since when the origin is off-center that may actually be 0 and thus
    // nothing gets searched ev 9/12/07
    private def computePatchRange(
      wrappingAllowed: Boolean,
      worldDimension: Int,
      minDimension: Int,
      maxDimension: Int, start: Int, r: Int): (Int, Int) = {
        if (wrappingAllowed) {
          val dimension = worldDimension / 2.0
          if (r < dimension) (r, -r)
          else (StrictMath.floor(dimension).toInt, - StrictMath.ceil(dimension - 1).toInt)
        } else {
          val diff = minDimension - start
          (StrictMath.min(maxDimension - start, r), if (StrictMath.abs(diff) < r) diff else -r)
        }
      }
  }

  private class RadialPatchSetEnumeration[A <: Agent](geometry: VerifyGeometry[A], membership: VerifyMembership[A],
    radius: Double, sourceSet: AgentSet, startPatch: Patch3D)(implicit ra: ThreeDRadialAgent[A])
    extends PatchSetEnumeration(geometry, membership, radius, sourceSet, startPatch) {
    def getPatch(dx: Int, dy: Int, dz: Int): Patch3D = startPatch.getPatchAtOffsets(dx, dy, dz)
  }

  private class ConicPatchSetEnumeration[A <: Agent](geometry: VerifyGeometry[A], membership: VerifyMembership[A],
    radius: Double, sourceSet: AgentSet, startPatch: Patch3D, world: World3D)(implicit ra: ThreeDRadialAgent[A])
    extends PatchSetEnumeration(geometry, membership, radius, sourceSet, startPatch) {
    val startPxcor = startPatch.pxcor
    val startPycor = startPatch.pycor
    val startPzcor = startPatch.pzcor
    def getPatch(dx: Int, dy: Int, dz: Int): Patch3D =
      world.getPatchAtWrap(startPxcor + dx, startPycor + dy, startPzcor + dz).asInstanceOf[Patch3D]
  }
}

File: netlogo-core/src/main/agent/InRadiusOrCone3D.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent;

import org.nlogo.api.AgentException;
import org.nlogo.api.Vect;
import org.nlogo.core.AgentKindJ;

import java.util.ArrayList;
import java.util.List;

public strictfp class InRadiusOrCone3D
  implements World.InRadiusOrCone {
  private final World3D world;

  InRadiusOrCone3D(World3D world) {
    this.world = world;
  }

  public List<Agent> inRadiusSimple(Agent agent, AgentSet sourceSet,
      double radius, boolean wrap) {
    return inRadius(agent, sourceSet, radius, wrap);
  }

  @Override
  public List<Agent> inRadius(Agent agent, AgentSet sourceSet,
                              double radius, boolean wrap) {
    int worldWidth = world.worldWidth();
    int worldHeight = world.worldHeight();
    int worldDepth = world.worldDepth();
    int maxPxcor = world.maxPxcor();
    int maxPycor = world.maxPycor();
    int minPxcor = world.minPxcor();
    int minPycor = world.minPycor();

    List<Agent> result = new ArrayList<Agent>();
    Patch3D startPatch;
    double startX, startY, startZ;
    if (agent instanceof Turtle) {
      Turtle3D startTurtle = (Turtle3D) agent;
      startPatch = (Patch3D) startTurtle.getPatchHere();
      startX = startTurtle.xcor();
      startY = startTurtle.ycor();
      startZ = startTurtle.zcor();
    } else {
      startPatch = (Patch3D) agent;
      startX = startPatch.pxcor;
      startY = startPatch.pycor;
      startZ = startPatch.pzcor;
    }

    int dxmin = 0;
    int dxmax = 0;
    int dymin = 0;
    int dymax = 0;
    int dzmin = 0;
    int dzmax = 0;

    int r = (int) StrictMath.ceil(radius);

    if (world.wrappingAllowedInX()) {
      double width = worldWidth / 2.0;
      if (r < width) {
        dxmax = r;
        dxmin = -r;
      } else {
        dxmax = (int) StrictMath.floor(width);
        dxmin = -(int) StrictMath.ceil(width - 1);
      }
    } else {
      int xdiff = minPxcor - startPatch.pxcor;
      dxmin = StrictMath.abs(xdiff) < r ? xdiff : -r;
      dxmax = StrictMath.min((maxPxcor - startPatch.pxcor), r);
    }
    if (world.wrappingAllowedInY()) {
      double height = worldHeight / 2.0;
      if (r < height) {
        dymax = r;
        dymin = -r;
      } else {
        dymax = (int) StrictMath.floor(height);
        dymin = -(int) StrictMath.ceil(height - 1);
      }
    } else {
      int ydiff = minPycor - startPatch.pycor;
      dymin = StrictMath.abs(ydiff) < r ? ydiff : -r;
      dymax = StrictMath.min((maxPycor - startPatch.pycor), r);
    }

    double depth = worldDepth / 2.0;
    if (r < depth) {
      dzmax = r;
      dzmin = -r;
    } else {
      dzmax = (int) StrictMath.floor(depth);
      dzmin = -(int) StrictMath.ceil(depth - 1);
    }

    Protractor3D protractor = world.protractor();

    for (int dz = dzmin; dz <= dzmax; dz++) {
      for (int dy = dymin; dy <= dymax; dy++) {
        for (int dx = dxmin; dx <= dxmax; dx++) {
          Patch3D patch = null;
          try {
            patch = startPatch.getPatchAtOffsets(dx, dy, dz);
          } catch (AgentException ex) {
            // at present wrapping is always on in 3D, so this can
            // never happen - ST 7/18/06
            throw new IllegalStateException(ex);
          }
          if (sourceSet.kind() == AgentKindJ.Patch()) {
            if (protractor.distance(patch.pxcor, patch.pycor, patch.pzcor,
                startX, startY, startZ,
                wrap)
                <= radius && (sourceSet == world.patches() || sourceSet.contains(patch))) {
              result.add(patch);
            }
          } else {
            // Only check patches that might have turtles within the radius on them.
            // The 1.415 (square root of 2) adjustment is necessary because it is
            // possible for portions of a patch to be within the circle even though
            // the center of the patch is outside the circle.
            if (StrictMath.sqrt(dx * dx + dy * dy + dz * dz) <= radius + 1.415) {
              for (Turtle turtle : patch.turtlesHere()) {
                if ((sourceSet == world.turtles() ||
                    // any turtle set with a non-null print name is either
                    // the set of all turtles, or a breed agentset - ST 2/19/04
                    (sourceSet.isBreedSet() && sourceSet == turtle.getBreed()) ||
                    (!sourceSet.isBreedSet() && sourceSet.contains(turtle))) &&
                    (protractor.distance(turtle.xcor(), turtle.ycor(),
                        ((Turtle3D) turtle).zcor(),
                        startX, startY, startZ, wrap)
                        <= radius)) {
                  result.add(turtle);
                }
              }
            }
          }
        }
      }
    }
    return result;
  }

  @Override
  public List<Agent> inCone(Turtle callingTurtle, AgentSet sourceSet,
                            double radius, double angle, boolean wrap) {
    int worldWidth = world.worldWidth();
    int worldHeight = world.worldHeight();
    int worldDepth = world.worldDepth();
    int maxPxcor = world.maxPxcor();
    int maxPycor = world.maxPycor();
    int minPxcor = world.minPxcor();
    int minPycor = world.minPycor();

    int m = 0;
    int n = 0;
    int k = 0;

    Turtle3D startTurtle = (Turtle3D) callingTurtle;

    if (wrap) {
      m = world.wrappingAllowedInX() ? (int) StrictMath.ceil(radius / worldWidth) : 0;
      n = world.wrappingAllowedInY() ? (int) StrictMath.ceil(radius / worldHeight) : 0;
      k = world.wrappingAllowedInZ() ? (int) StrictMath.ceil(radius / worldDepth) : 0;
    }

    List<Agent> result = new ArrayList<Agent>();
    Patch3D startPatch = (Patch3D) startTurtle.getPatchHere();
    double half = angle / 2;
    // dxmax and dymax determine which patches we will check.  the patches we
    // check always form a rectangle.  usually it will be a square, but if
    // the radius is large enough, then it might be a rectangle.

    int dxmin = 0;
    int dxmax = 0;
    int dymin = 0;
    int dymax = 0;
    int dzmin = 0;
    int dzmax = 0;

    int r = (int) StrictMath.ceil(radius);

    if (world.wrappingAllowedInX()) {
      double width = worldWidth / 2.0;
      if (r < width) {
        dxmax = r;
        dxmin = -r;
      } else {
        dxmax = (int) StrictMath.floor(width);
        dxmin = -(int) StrictMath.ceil(width - 1);
      }
    } else {
      int xdiff = minPxcor - startPatch.pxcor;
      dxmin = StrictMath.abs(xdiff) < r ? xdiff : -r;
      dxmax = StrictMath.min((maxPxcor - startPatch.pxcor), r);
    }
    if (world.wrappingAllowedInY()) {
      double height = worldHeight / 2.0;
      if (r < height) {
        dymax = r;
        dymin = -r;
      } else {
        dymax = (int) StrictMath.floor(height);
        dymin = -(int) StrictMath.ceil(height - 1);
      }
    } else {
      int ydiff = minPycor - startPatch.pycor;
      dymin = StrictMath.abs(ydiff) < r ? ydiff : -r;
      dymax = StrictMath.min((maxPycor - startPatch.pycor), r);
    }

    double depth = worldDepth / 2.0;
    if (r < depth) {
      dzmax = r;
      dzmin = -r;
    } else {
      dzmax = (int) StrictMath.floor(depth);
      dzmin = -(int) StrictMath.ceil(depth - 1);
    }

    // loop through the patches in the rectangle.  (it doesn't matter what
    // order we check them in.)
    for (int dz = dzmin; dz <= dzmax; dz++) {
      for (int dy = dymin; dy <= dymax; dy++) {
        for (int dx = dxmin; dx <= dxmax; dx++) {
          Patch3D patch = (Patch3D) world.getPatchAtWrap
              (startPatch.pxcor + dx, startPatch.pycor + dy, startPatch.pzcor + dz);

          if (patch != null) {
            if (sourceSet.kind() == AgentKindJ.Patch()) {
              // loop through our world copies
              outer:
              for (int worldOffsetX = -m; worldOffsetX <= m; worldOffsetX++) {
                for (int worldOffsetY = -n; worldOffsetY <= n; worldOffsetY++) {
                  for (int worldOffsetZ = -k; worldOffsetZ <= k; worldOffsetZ++) {
                    if ((sourceSet == world.patches() || sourceSet.contains(patch))
                        && isInCone(patch.pxcor + worldWidth * worldOffsetX,
                        patch.pycor + worldHeight * worldOffsetY,
                        patch.pzcor + worldDepth * worldOffsetZ,
                        startTurtle.xcor(), startTurtle.ycor(),
                        startTurtle.zcor(),
                        radius, half, startTurtle.heading(),
                        startTurtle.pitch())) {
                      result.add(patch);
                      break outer;
                    }
                  }
                }
              }
            } else {
              if (StrictMath.sqrt(dx * dx + dy * dy + dz * dz) <= radius + 1.415) {
                for (Turtle turtle : patch.turtlesHere()) {
                  // loop through our world copies
                  outer:
                  for (int worldOffsetX = -m; worldOffsetX <= m; worldOffsetX++) {
                    for (int worldOffsetY = -n; worldOffsetY <= n; worldOffsetY++) {
                      for (int worldOffsetZ = -k; worldOffsetZ <= k; worldOffsetZ++) {
                        // any turtle set with a non-null print name is either
                        // the set of all turtles, or a breed agentset - ST 2/19/04
                        if ((sourceSet == world.turtles() ||
                            (sourceSet.isBreedSet() && sourceSet == turtle.getBreed()) ||
                            (!sourceSet.isBreedSet() && sourceSet.contains(turtle))) &&
                            isInCone(turtle.xcor() + worldWidth * worldOffsetX,
                                turtle.ycor() + worldHeight * worldOffsetY,
                                ((Turtle3D) turtle).zcor() + worldDepth * worldOffsetZ,
                                startTurtle.xcor(), startTurtle.ycor(), startTurtle.zcor(),
                                radius, half, startTurtle.heading(),
                                startTurtle.pitch())) {
                          result.add(turtle);
                          break outer;
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
    return result;
  }

  private boolean isInCone(double x, double y, double z,
                           double cx, double cy, double cz,
                           double r, double half, double h, double p) {
    Protractor3D protractor = world.protractor();

    if (x == cx && y == cy && z == cz) {
      return true;
    }
    if (protractor.distance(cx, cy, cz, x, y, z, false) > r) // false = don't wrap, since inCone()
    // handles wrapping its own way
    {
      return false;
    }

    Vect unitVect = Vect.toVectors(h, p, 0)[0];
    Vect targetVect = new Vect(x - cx, y - cy, z - cz);
    double angle = targetVect.angleTo(unitVect);
    double halfRadians = StrictMath.toRadians(half);

    return angle <= halfRadians || angle >= (2 * StrictMath.PI - halfRadians);
  }
}
